### PR TITLE
fix: [DEL-4515]: Parallelising secret evaluation in cg manager

### PR DESCRIPTION
**Pull request by gaurav-jain1**
## Describe your changes
Changes to parllelize secret evaluation in Manager. This will be behind ff, so that can turned on/off when required.
## Checklist
- [x] I've documented the changes in the PR description.
- [ ] I've tested this change either in PR or local environment.
- [ ] If hashcheck failed I followed [the checklist](https://harness.atlassian.net/wiki/spaces/DEL/pages/21016838831/PR+Codebasehash+Check+merge+checklist) and updated this PR description with my answers to make sure I'm not introducing any breaking changes.

## Comment Triggers
<details>
  <summary>Build triggers</summary>
  
- Feature build: `trigger feature-build`
- Immutable delegate `trigger publish-delegate`
</details>

<details>
  <summary>PR Check triggers</summary>
  
  You can run multiple PR check triggers by comma separating them in a single comment. e.g. `trigger ti0, ti1`
  
- Compile: `trigger compile`
- CodeFormat: `trigger codeformat`
- MessageMetadata: `trigger messagecheck`
- Recency: `trigger recency`
- BuildNumberMetadata: `trigger buildnum`
- runDockerizationCheck: `trigger dockerizationcheck`
- runAuthorCheck: `trigger authorcheck`
- Checkstyle: `trigger checkstyle`
- PMD: `trigger pmd`
- TI-bootstrap: `trigger ti0`
- TI-bootstrap1: `trigger ti1`
- TI-bootstrap2: `trigger ti2`
- TI-bootstrap3: `trigger ti3`
- TI-bootstrap4: `trigger ti4`
- FunctionalTest1: `trigger ft1`
- FunctionalTest2: `trigger ft2`
- CodeBaseHash: `trigger codebasehash`
</details>

## PR check failures and solutions
https://harness.atlassian.net/wiki/spaces/BT/pages/21106884744/PR+Checks+-+Failures+and+Solutions

## [Contributor license agreement](https://github.com/harness/harness-core/blob/develop/CONTRIBUTOR_LICENSE_AGREEMENT.md)

- - -
This change is [Reviewable](https://reviewable.io/reviews/harness/harness-core/37441)

### CODE REVIEW

**Comment by diptiman-adak** [comment]
Quality gate Quality gate: **_Failed_** 
 


            
 #### 9 Issues 
* Vulnerability 2 Vulnerabilities 
 * Code issue 6 Code issues 
* Anti-pattern 1 Anti-patterns 

 [Details](https://app.embold.io/organization/gh/harness/repositories/7752dbb3099d2dba3b5d37f5931a5e4c/pull-request/37441)

File: 400-rest/src/main/java/software/wings/expression/NgSecretManagerFunctor.java [hunk 1]
```diff
@@ -83,6 +88,13 @@ public Object obtain(String secretIdentifier, int token) {
       throw new FunctorException("Inappropriate usage of internal functor");
     }
     try {
+      if (!evaluateSync) {
+        if (expressionEvaluatorExecutor != null) {
+          // Offload expression evaluation of secrets to another threadpool.
+          return expressionEvaluatorExecutor.submit(() -> obtainInternal(secretIdentifier));
```
Comment (abhinav-harness) [inline 400-rest/src/main/java/software/wings/expression/NgSecretManagerFunctor.java:94]: this was returning evaluated object now the future

Comment (gaurav-jain1) [reply 400-rest/src/main/java/software/wings/expression/NgSecretManagerFunctor.java:94]: Yes, that's how we are making it async

**Comment by gaurav-jain1** [comment]
trigger ti2

File: 980-commons/src/main/java/io/harness/expression/ImageSecretFunctor.java [hunk 2]
```diff
@@ -27,4 +29,15 @@ public String create(String registryUrl, String userName, String password) {
     return EncodingUtils.encodeBase64(format(
         REGISTRY_CREDENTIAL_TEMPLATE, registryUrl, userName, password.replaceAll("\n", "").replaceAll("\"", "\\\\\"")));
   }
+
+  // This variation will be used in case of nested secrets, which has not been evaluated yet.
+  public String create(String registryUrl, String userName, Future password) {
+    String evaluatedPassword;
```
Comment (abhinav-harness) [inline 980-commons/src/main/java/io/harness/expression/ImageSecretFunctor.java:35]: how is this functor used ?.

**Comment by raghuAtWings** [comment]
*[`400-rest/src/main/java/software/wings/expression/ManagerPreExecutionExpressionEvaluator.java` line 77 at r1](https://reviewable.io/reviews/harness/harness-core/37441#-NBy0YAq2TPV9Dt4-uMm:-NBy0YAq2TPV9Dt4-uMn:b-fer6fj) ([raw file](https://github.com/harness/harness-core/blob/def7b5018616f46700316b5d365bea13ad0ee5bf/400-rest/src/main/java/software/wings/expression/ManagerPreExecutionExpressionEvaluator.java#L77)):*
> ```Java
>             .build());
> 
>     final boolean evalExpInSync = featureFlagService.isEnabled(FeatureName.EVALUATE_SECRET_EXPRESSION_SYNC, accountId);
> ```

do we need FF? what happens if this misbehaves ?

**Comment by raghuAtWings** [comment]
*[`400-rest/src/main/java/software/wings/expression/NgSecretManagerFunctor.java` line 91 at r1](https://reviewable.io/reviews/harness/harness-core/37441#-NBy1e7XBq9SrpgBF4WK:-NBy1e7XBq9SrpgBF4WL:b-lovzz5) ([raw file](https://github.com/harness/harness-core/blob/def7b5018616f46700316b5d365bea13ad0ee5bf/400-rest/src/main/java/software/wings/expression/NgSecretManagerFunctor.java#L91)):*
> ```Java
>     }
>     try {
>       if (!evaluateSync) {
> ```

so we added a FF and we are doing negation here, so by default everyone will use the threads to decrypt but we can disable by enabling the flag?

**Comment by raghuAtWings** [comment]
*[`400-rest/src/main/java/software/wings/expression/NgSecretManagerFunctor.java` line 91 at r1](https://reviewable.io/reviews/harness/harness-core/37441#-NBy1e7XBq9SrpgBF4WK:-NBy2BZ89EwE9LM862DS:bh6chnl) ([raw file](https://github.com/harness/harness-core/blob/def7b5018616f46700316b5d365bea13ad0ee5bf/400-rest/src/main/java/software/wings/expression/NgSecretManagerFunctor.java#L91)):*
<details><summary><i>Previously, raghuAtWings (Raghvendra Singh) wrote…</i></summary><blockquote>

so we added a FF and we are doing negation here, so by default everyone will use the threads to decrypt but we can disable by enabling the flag?
</blockquote></details>

also check the FF naming guide at https://harness.atlassian.net/wiki/spaces/SPG/pages/21183823964/Feature+Flag+Guidelines

**Review by abhinav-harness — APPROVED** [review]
> Reviewed 6 of 9 files at r1, 7 of 7 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/37441)* status: :shipit: complete! ✓

**Review by raghuAtWings — DISMISSED** [review]
> Reviewed 6 of 9 files at r1, 5 of 7 files at r2, 1 of 1 files at r3, 1 of 1 files at r4, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/37441)* status: :shipit: complete! ✓

**Review by abhinav-harness — APPROVED** [review]
> Reviewed 1 of 1 files at r3, 1 of 1 files at r5, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/37441)* status: :shipit: complete! ✓

**Pull request merged**